bftengine/tests/bcstatetransfer/source_selector_test.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+TEST(source_selector_test, reset_sets_selection_time_to_zero) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.isReset());
+  source_selector.setSourceSelectionTime(kSampleCurrentTimeMs);
+  ASSERT_FALSE(source_selector.isReset());
+  source_selector.reset();
+  ASSERT_TRUE(source_selector.isReset());
+}
+
+TEST(source_selector_test, reset_sets_the_fetching_timestamp_to_zero) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.isReset());
+  source_selector.setFetchingTimeStamp(GL, kSampleCurrentTimeMs);
+  ASSERT_FALSE(source_selector.isReset());
+  source_selector.reset();
+  ASSERT_TRUE(source_selector.isReset());
+}
+
+// Others
+TEST(source_selector_test, set_all_replicas_as_preferred) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_FALSE(source_selector.hasPreferredReplicas());
+  source_selector.setAllReplicasAsPreferred();
+  ASSERT_TRUE(source_selector.hasPreferredReplicas());
+
+  for (const auto& replica : replicas) {
+    ASSERT_TRUE(source_selector.isPreferred(replica));
+  }
+}
+
+TEST(source_selector_test, select_an_unique_replica_on_each_selection) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  std::set<uint16_t> selected_replicas;
+
+  // Select an initial replica
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  ASSERT_TRUE(selected_replicas.insert(source_selector.currentReplica()).second);
+
+  // Loop until the only preferred replica left is the one that has been selected initially
+  while (source_selector.numberOfPreferredReplicas() > 1) {
+    source_selector.updateSource(kSampleCurrentTimeMs);
+    ASSERT_TRUE(selected_replicas.insert(source_selector.currentReplica()).second);
+  }
+
+  ASSERT_EQ(replicas, selected_replicas);
+}
+
+// Operating with an empty set of replicas
+TEST(source_selector_test, cannot_select_source_without_initial_replicas) {
+  auto source_selector = SourceSelector({}, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_DEATH(source_selector.updateSource(kSampleCurrentTimeMs), "");
+}
+
+TEST(source_selector_test, add_preferred_replica) {
+  auto source_selector = SourceSelector({}, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  source_selector.addPreferredReplica(kSampleReplicaId);
+  ASSERT_TRUE(source_selector.isPreferred(kSampleReplicaId));
+}
+
+TEST(source_selector_test, select_the_only_preferred_replica) {
+  auto source_selector = SourceSelector({}, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  source_selector.addPreferredReplica(kSampleReplicaId);
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  ASSERT_TRUE(source_selector.isPreferred(kSampleReplicaId) && source_selector.currentReplica() == kSampleReplicaId);
```

Comment:
I think separate asserts would be better.

Reply:
The asserts have been separated.